enhancements/sig-architecture/149-moving-cluster-proxy-to-core/README.md

```diff
@@ -0,0 +1,167 @@
+
+Currently, `cluster-proxy` is implemented based on the ANP (APIServer Network Proxy) server and deployed as an addon.
+However, we have identified several limitations when using `cluster-proxy`:
+
+1. `cluster-proxy` provides a `kconnectivity` interface which is gRPC-based. It is difficult to call directly
+   with Kubernetes clients, requiring an HTTP proxy frontend to handle requests.
+2. `cluster-proxy` leverages the ANP server which transports TCP packets over gRPC tunnels. The agent directly
+   proxies TCP packets to the target location, making it difficult for the agent to act as a proxy or delegator
+   to mutate and proxy HTTP requests.
+3. End users must always enable the cluster-proxy addon, introducing additional operational overhead when
+   cluster-proxy functionality is commonly needed.
+
+We are introducing gRPC as a registration mechanism in OCM, which makes it easier to move cluster-proxy
+capabilities into the klusterlet.
+
+### Goals
+
+- Add a feature flag `ClusterProxy` in both klusterlet and clustermanager. When enabled, a gRPC server will
+  start on the hub, and the klusterlet agent will connect to the gRPC server for proxy functionality.
+- When the feature is enabled, ClusterManager starts an HTTP server to proxy requests to the target cluster
+  via the klusterlet agent.
+- All existing capabilities of `cluster-proxy` will be preserved.
+
+### Non-Goals
+
+- How to deprecate the `cluster-proxy addon` is not in the scope of this proposal, given that `cluster-proxy`
+  is still consumed by `MulticlusterGateway` and other consumers.
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+Users can enable the `ClusterProxy` feature gate in both clustermanager and klusterlet. Once enabled, users can
+use the HTTP server on the hub cluster to proxy requests to the API server of the targeted cluster.
+
+#### Story 2
+
+Users can authenticate via proxy using either the token for the targeted cluster, or the token for the hub
+cluster if impersonation permissions are granted to the klusterlet agent.
+
+
+### Risks and Mitigation
+
+N/A
+
+## Design Details
+
+#### Registration
+The connection between klusterlet and hub for proxy functionality will remain gRPC-based. This is independent
+of the gRPC registration driver, requiring a separate gRPC configuration file for the proxy. To obtain the gRPC
+configuration for proxy, the klusterlet agent will send a Certificate Signing Request (CSR) to the hub with the
+signer name `open-cluster-management.io/klusterlet-proxy`. The gRPC server on the hub will approve and sign the
+CSR. The gRPC configuration will be saved in the `hub-kubeconfig-secret` with the key `proxy-grpc.yaml`. The
+klusterlet will only start the proxy when it detects the existence of this configuration file. The proxy
+registration process begins after the cluster registration process and will not impact normal cluster
+registration.
+
+#### Proxy API
```

Comment:
I wonder whether we should use term "proxy", I got same case users get confused about cluster-proxy with other "proxy" settings like addonproxyconfig.proxyconfig.

Could we use another term like "tunnel"?

Reply:
I believe the complete description of this functionality can be "proxy requests between hub and agent," but it can also be "establish a tunnel between hub and agent that can be used to proxy requests."

For users of this functionality, they indeed only need to know about an endpoint that can proxy requests - they don't need to know about the existence of a "tunnel." However, for developers and maintainers, the term "tunnel" helps distinguish this functionality from other proxy-related features.

If "tunnel" doesn't work, could we adopt other terms such as: gateway, bridge, relay, etc.?

Reply:
relay seems ok to me, but since cluster-proxy has been known to many, I am not quite sure how strong the desire to rename. 

@jnpacker thought?